<a href="https://colab.research.google.com/github/exwaized/aaditya/blob/main/sonarr.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Task
Create a mine vs rock predictor machine learning program that takes input from the Excel file "Sonar_Mines_vs_Rocks.xlsx" and outputs if it is rock or mine.

## Load data

### Subtask:
Load the data from the Excel file into a pandas DataFrame.


**Reasoning**:
Import pandas and load the data from the Excel file into a DataFrame, then display the head and info to inspect the data.



In [10]:
import pandas as pd

# Load the data from the CSV file
df = pd.read_csv("Copy of sonar data.csv")

# Display the head and info to inspect the data
display(df.head())
display(df.info())

0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109  \
0  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
1  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
2  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
3  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
4  0.0286  0.0453  0.0277  0.0174  0.0384  0.0990  0.1201  0.1833  0.2105   

   0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084  \
0  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
1  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
2  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
3  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   
4  0.3039  ...  0.0045  0.0014  0.0038  0.0013  0.0089  0.0057  0.0027   

   0.0090  0.0032  R  
0  0.0052  0.0044  R  
1  0.0095  0.0078  R  
2  0.0040  0.0117  R  
3  0.0107  0.0094  R  
4  0.0051  0.0062  R  

[5 rows x 61 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207 entries, 0 to 206
Data columns (total 61 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0.0200  207 non-null    float64
 1   0.0371  207 non-null    float64
 2   0.0428  207 non-null    float64
 3   0.0207  207 non-null    float64
 4   0.0954  207 non-null    float64
 5   0.0986  207 non-null    float64
 6   0.1539  207 non-null    float64
 7   0.1601  207 non-null    float64
 8   0.3109  207 non-null    float64
 9   0.2111  207 non-null    float64
 10  0.1609  207 non-null    float64
 11  0.1582  207 non-null    float64
 12  0.2238  207 non-null    float64
 13  0.0645  207 non-null    float64
 14  0.0660  207 non-null    float64
 15  0.2273  207 non-null    float64
 16  0.3100  207 non-null    float64
 17  0.2999  207 non-null    float64
 18  0.5078  207 non-null    float64
 19  0.4797  207 non-null    float64
 20  0.5783  207 non-null    float64
 21  0.5071  207 non-null    float64
 22  0.

None

**Reasoning**:
The previous attempt failed because the file was not found. I will try to list the files in the current directory to verify the file name and location.



In [2]:
import os
print(os.listdir())

['.config', '.ipynb_checkpoints', 'Copy of sonar data.csv', 'sample_data']


**Reasoning**:
The file name was incorrect in the previous attempt. Based on the output of `os.listdir()`, the correct file name is "Copy of sonar data.csv". I will try to load this CSV file instead of the Excel file.



In [3]:
df = pd.read_csv("Copy of sonar data.csv")

display(df.head())
display(df.info())

0.0200  0.0371  0.0428  0.0207  0.0954  0.0986  0.1539  0.1601  0.3109  \
0  0.0453  0.0523  0.0843  0.0689  0.1183  0.2583  0.2156  0.3481  0.3337   
1  0.0262  0.0582  0.1099  0.1083  0.0974  0.2280  0.2431  0.3771  0.5598   
2  0.0100  0.0171  0.0623  0.0205  0.0205  0.0368  0.1098  0.1276  0.0598   
3  0.0762  0.0666  0.0481  0.0394  0.0590  0.0649  0.1209  0.2467  0.3564   
4  0.0286  0.0453  0.0277  0.0174  0.0384  0.0990  0.1201  0.1833  0.2105   

   0.2111  ...  0.0027  0.0065  0.0159  0.0072  0.0167  0.0180  0.0084  \
0  0.2872  ...  0.0084  0.0089  0.0048  0.0094  0.0191  0.0140  0.0049   
1  0.6194  ...  0.0232  0.0166  0.0095  0.0180  0.0244  0.0316  0.0164   
2  0.1264  ...  0.0121  0.0036  0.0150  0.0085  0.0073  0.0050  0.0044   
3  0.4459  ...  0.0031  0.0054  0.0105  0.0110  0.0015  0.0072  0.0048   
4  0.3039  ...  0.0045  0.0014  0.0038  0.0013  0.0089  0.0057  0.0027   

   0.0090  0.0032  R  
0  0.0052  0.0044  R  
1  0.0095  0.0078  R  
2  0.0040  0.0117  R  
3  0.0107  0.0094  R  
4  0.0051  0.0062  R  

[5 rows x 61 columns]

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 207 entries, 0 to 206
Data columns (total 61 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   0.0200  207 non-null    float64
 1   0.0371  207 non-null    float64
 2   0.0428  207 non-null    float64
 3   0.0207  207 non-null    float64
 4   0.0954  207 non-null    float64
 5   0.0986  207 non-null    float64
 6   0.1539  207 non-null    float64
 7   0.1601  207 non-null    float64
 8   0.3109  207 non-null    float64
 9   0.2111  207 non-null    float64
 10  0.1609  207 non-null    float64
 11  0.1582  207 non-null    float64
 12  0.2238  207 non-null    float64
 13  0.0645  207 non-null    float64
 14  0.0660  207 non-null    float64
 15  0.2273  207 non-null    float64
 16  0.3100  207 non-null    float64
 17  0.2999  207 non-null    float64
 18  0.5078  207 non-null    float64
 19  0.4797  207 non-null    float64
 20  0.5783  207 non-null    float64
 21  0.5071  207 non-null    float64
 22  0.

None

## Preprocess data

### Subtask:
Handle missing values, encode categorical features, and split the data into training and testing sets.


**Reasoning**:
Rename the last column, check for missing values, encode the target variable, and split the data into training and testing sets.



In [4]:
# Rename the last column to 'target'
df = df.rename(columns={df.columns[-1]: 'target'})

# Check for missing values
# print("Missing values before handling:")
# print(df.isnull().sum())
# Since there are no missing values, no handling is needed.

# Encode the target column
df['target'] = df['target'].astype('category').cat.codes

# Separate features and target
X = df.drop('target', axis=1)
y = df['target']

# Split the data into training and testing sets
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

# Print the shapes of the training and testing sets
# print("\nShape of X_train:", X_train.shape)
# print("Shape of X_test:", X_test.shape)
# print("Shape of y_train:", y_train.shape)
# print("Shape of y_test:", y_test.shape)

Missing values before handling:
0.0200    0
0.0371    0
0.0428    0
0.0207    0
0.0954    0
         ..
0.0180    0
0.0084    0
0.0090    0
0.0032    0
target    0
Length: 61, dtype: int64

Shape of X_train: (165, 60)
Shape of X_test: (42, 60)
Shape of y_train: (165,)
Shape of y_test: (42,)


## Train a model

### Subtask:
Choose a suitable classification model and train it on the training data.


**Reasoning**:
Import and train a Logistic Regression model on the training data.



In [5]:
from sklearn.linear_model import LogisticRegression

# Initialize and train a Logistic Regression model
model = LogisticRegression()
model.fit(X_train, y_train)

LogisticRegression()

## Evaluate the model

### Subtask:
Evaluate the trained model's performance on the testing data.


**Reasoning**:
Evaluate the performance of the trained model on the test set using accuracy and a classification report.



In [6]:
from sklearn.metrics import accuracy_score, classification_report

# Make predictions on the test set
y_pred = model.predict(X_test)

# Evaluate the model's performance
accuracy = accuracy_score(y_test, y_pred)
print(f"Accuracy: {accuracy:.2f}")

report = classification_report(y_test, y_pred)
print("Classification Report:\n", report)

Accuracy: 0.76
Classification Report:
               precision    recall  f1-score   support

           0       0.95      0.67      0.78        27
           1       0.61      0.93      0.74        15

    accuracy                           0.76        42
   macro avg       0.78      0.80      0.76        42
weighted avg       0.83      0.76      0.77        42



## Make predictions

### Subtask:
Use the trained model to make predictions on new data from the Excel file.


**Reasoning**:
Use the trained model to make predictions on the entire dataset (`X`) and store the predictions.



In [13]:
# Use the trained model to make predictions on the entire dataset
all_predictions = model.predict(X)

## Summary:

### Data Analysis Key Findings

*   The dataset was successfully loaded from the "Copy of sonar data.csv" file, containing 207 entries and 61 columns.
*   There were no missing values found in the dataset.
*   The target variable was successfully encoded into numerical values.
*   The data was split into training and testing sets, with 80% for training and 20% for testing. The training set has 165 samples and the testing set has 42 samples.
*   A Logistic Regression model was trained on the training data.
*   The trained model achieved an accuracy of 76% on the testing data.
*   The classification report showed that the model had high precision (0.95) and lower recall (0.67) for class 0 (likely "rock"), and lower precision (0.61) and high recall (0.93) for class 1 (likely "mine").
*   Predictions were successfully made on the entire dataset using the trained model.

### Insights or Next Steps

*   Consider exploring other classification models (e.g., Support Vector Machines, Random Forests) to potentially improve accuracy.
*   Investigate the imbalance in precision and recall between the two classes to understand if there are underlying data characteristics or model biases contributing to this difference.


In [12]:
import numpy as np

# Take the first row of the DataFrame as input
user_input_array = X.iloc[0].values.reshape(1, -1)

# Make a prediction using the trained model
prediction = model.predict(user_input_array)

# Decode the prediction (0: Mine, 1: Rock)
decoded_prediction = 'Rock' if prediction[0] == 1 else 'Mine'

print(f"\nThe prediction for the first row is: {decoded_prediction}")


The prediction for the first row is: Rock


/usr/local/lib/python3.11/dist-packages/sklearn/utils/validation.py:2739: UserWarning: X does not have valid feature names, but LogisticRegression was fitted with feature names
  warnings.warn(
